lms/services/canvas.py

```diff
@@ -1,36 +1,79 @@
-                raise CanvasFileNotFoundInCourse(file_id)
+        self._file_service = file_service
+
+    def public_url_for_file(
+        self, module_item_configuration, file_id, course_id, check_in_course=False
+    ):
+        mapped_file_id = module_item_configuration.get_canvas_mapped_file_id(file_id)
+
+        # If there's a previously stored mapping for file_id use that instead.
+        effective_file_id = mapped_file_id or file_id
+
+        try:
+            return self._public_url(
+                effective_file_id,
+                course_id=course_id if check_in_course else None,
+            )
+        except (CanvasFileNotFoundInCourse, CanvasAPIPermissionError):
+            # Either the user can't see the file in the current course's list
+            # of files or the user got a permissions error from the Canvas API
+            # when trying to get a public URL for the file.
+            #
+            # This can happen because the course has been copied and the
+            # assignment's file_id is from the original course. Or it can
+            # happen because the assignment's file has been deleted from
+            # Canvas. Or it can happen because the user doesn't have permission
+            # to see the file ("unpublished" files in Canvas are visible to
+            # instructors but not to students).
+            #
+            # We'll try to find another copy of the same file that the current
+            # user *can* see in the current course and use that instead.
+
+            # Look for a previously saved record of the assignment's file in our DB.
+            file_ = self._file_service.get(effective_file_id, type_="canvas_file")
+
+            if not file_:
+                # We don't have a record of the assignment's file in our DB.
+                # This can happen, for example, if Hypothesis has not been
+                # launched in the original assignment's course since we
+                # deployed the code that started recording files in the DB.
+                raise
+
+            # Look for a copy of the assignment's file that the current user
+            # *can* see in the current course.
+            found_file_id = self.find_matching_file_in_course(course_id, file_)
+
+            if not found_file_id:
+                # We didn't find a matching file in the current course.
+                # This could mean that the file has been deleted, has been
+                # renamed, or the current user doesn't have permission to see
+                # the file.
+                raise
+
+            # We found a matching copy of the assignment's file that the
+            # current user *can* see in the current course. Store a mapping so
+            # that we don't have to re-do the search the next time the
+            # assignment is launched.
+            module_item_configuration.set_canvas_mapped_file_id(file_id, found_file_id)
```

Comment:
```suggestion
            module_item_configuration.set_mapped_file_id(file_id, found_file_id, type_="canvas_file")
```

Reply:
I'm gonna do this in a separate PR since `ModuleItemConfiguration.{get,set}_canvas_mapped_file_id()` aren't part of this PR.

How about we just drop the `type_` and just call them "file ID mappings"? A single `ModuleItemConfiguration` can't belong to more than one different LMS so I don't think there's a chance of a conflict:

```python
def get_mapped_file_id(self, file_id):
    return self.extra.get("file_mappings", {}).get(file_id)

def set_mapped_file_id(self, file_id, mapped_file_id):
    self.extra.setdefault("file_mappings", {})[file_id] = mapped_file_id
```

Reply:
https://github.com/hypothesis/lms/pull/2913